### PR TITLE
Improve autosave and backup feature

### PR DESCRIPTION
**Pull request by obraliar**
Applies the remaining feedback from https://github.com/JabRef/jabref/pull/2118 and fixes https://github.com/JabRef/jabref/issues/2192.
- [x] Manually tested changed features in running JabRef
- [x] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [x] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

File: src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java [hunk 1]
```diff
@@ -32,6 +32,8 @@
 /**
  * Backups the given bib database file from {@link BibDatabaseContext} on every {@link BibDatabaseContextChangedEvent}.
  * An intelligent {@link ExecutorService} with a {@link BlockingQueue} prevents a high load while making backups and rejects all redundant backup tasks.
+ *
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java]: Why is the space here?

File: src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java [hunk 2]
```diff
@@ -72,8 +74,13 @@ public void run() {
         }
     };
 
+
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java]: Why adding new empty line?

File: src/main/java/net/sf/jabref/logic/exporter/BibDatabaseWriter.java [hunk 3]
```diff
@@ -158,7 +158,11 @@ public E savePartOfDatabase(BibDatabaseContext bibDatabaseContext,
 
         session = saveSessionFactory.createSaveSession(preferences.getEncodingOrDefault(), preferences.getMakeBackup());
 
-        writeDatabaseID(bibDatabaseContext.getDatabase().getDatabaseID());
+        Optional<String> sharedDatabaseIDOptional = bibDatabaseContext.getDatabase().getSharedDatabaseID();
+
+        if (sharedDatabaseIDOptional.isPresent()) {
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/logic/exporter/BibDatabaseWriter.java]: why not use ifPresent?

File: src/main/java/net/sf/jabref/model/database/BibDatabase.java [hunk 4]
```diff
@@ -586,15 +588,33 @@ private void relayEntryChangeEvent(FieldChangedEvent event) {
...
+        return Optional.ofNullable(this.sharedDatabaseID);
+    }
+
+    public boolean isShared() {
+        return getSharedDatabaseID().isPresent();
     }
 
     /**
-     * @param databaseID use null to unset the id
+     * @param sharedDatabaseID use null to unset the id
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/model/database/BibDatabase.java]: I thought null shouldn't be used? Shouldn't the clear method used in this case?

File: src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java (hunk 1 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java]: Done.

File: src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java (hunk 2 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/logic/autosaveandbackup/BackupManager.java]: Done.

File: src/main/java/net/sf/jabref/logic/exporter/BibDatabaseWriter.java (hunk 3 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/logic/exporter/BibDatabaseWriter.java]: Not recommended due to Exception handling. `try`-`catch` would take more lines.

Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/logic/exporter/BibDatabaseWriter.java]: you could use orElseThrow() (http://stackoverflow.com/questions/28596790/throw-an-exception-if-an-optional-is-present). But from my side it's ok how it is.

**Comment by koppor** [comment]
[LGTM](https://lgtm.in/i/cghBUT10q) (LGTM)

File: src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java [hunk 5]
```diff
@@ -125,8 +125,7 @@ public void testGetDescription() {
     @Test
     public void testRecognizesDatabaseID() throws Exception {
         Path file = Paths.get(BibtexImporterTest.class.getResource("AutosavedSharedDatabase.bib").toURI());
-        String databaseID = importer.importDatabase(file, StandardCharsets.UTF_8).getDatabase().getDatabaseID();
-        assertEquals("13ceoc8dm42f5g1iitao3dj2ap", databaseID);
+        String sharedDatabaseID = importer.importDatabase(file, StandardCharsets.UTF_8).getDatabase().getSharedDatabaseID().get();
```
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java]: Compare to Optional.of(...) (also below)

Comment (obraliar) [inline src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java]: Done.

**Comment by obraliar** [comment]
@tobiasdiez 
> Can you please also add a test that the comment at the beginning of the file with the shared database id is not taken as the comment of the first bibentry.

Done. But what's the sense of this test? `%` marks up a comment generally.

File: src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java [hunk 6]
```diff
@@ -125,7 +126,17 @@ public void testGetDescription() {
...
+    }
+
+    @Test
+    public void testDatabaseIDisNotABibEntryComment() throws Exception {
+        Path file = Paths.get(BibtexImporterTest.class.getResource("AutosavedSharedDatabase.bib").toURI());
+        BibDatabase database = importer.importDatabase(file, StandardCharsets.UTF_8).getDatabase();
+
+        BibEntry entry = database.getEntries().get(0);
+
+        assertTrue(!entry.getField("comment").isPresent());
```
Comment (obraliar) [inline src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java]: Done.

**Comment by tobiasdiez** [comment]
BibTeX does not know the concept of "%"-comments. Everything outside of a BibEntry is considered a comment and is simply ignored by bibtex. However, JabRef safes the text before an entry as "user comments" so that they are kept upon rewriting an entry. However, we don't want to store the shared-db-id as a usercomment, since its an internal JabRef thing.

Please change the test code to "assertEquals("", string)"

**Comment by obraliar** [comment]
@tobiasdiez Done. Also thanks for the explantation.

**Comment by tobiasdiez** [comment]
One last question before merge: does the test fails if you remove the encoding line? (I would prefer if you write the tests against strings as opposed to bib-files)

**Comment by obraliar** [comment]
Yes, it fails...

**Comment by koppor** [comment]
Does that form a second test case? 😇

**Comment by obraliar** [comment]
No, the first one fails. The second one does not check whether a Shared ID is present. I could also add this assert to the second test but its already present in the previous one. Should I?

**Comment by tobiasdiez** [comment]
I think the cleanest way is to add a second test, which makes sure that a single db-string in front of an entry is not captured as "user comment". And then fix the code so that the test passes. 
(please move the tests to the ParserTests and don't use bib files as input but strings)

**Comment by koppor** [comment]
Other than my minor comment: LGTM. @tobiasdiez It seems, your test is now also included and working.

File: src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java [hunk 7]
```diff
@@ -121,11 +121,4 @@ public void testGetDescription() {
                         + "The metadata is not required to be read here, as this class is NOT called at --import.",
                 importer.getDescription());
     }
-
-    @Test
-    public void testRecognizesDatabaseID() throws Exception {
```
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java]: and please re-add this test in the parser test class.

**Comment by tobiasdiez** [comment]
Yes, also looks good to me. If the `testRecognizesDatabaseID` is added, the PR can be merged.

File: src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java (hunk 7 above)
Comment (obraliar) [inline src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java]: @koppor A test file has to be used as `importer.importDatabase(...)` requires a `File`. @tobiasdiez There was already a testing method `parseRecognizesDatabaseID()`.

Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/logic/importer/fileformat/BibtexImporterTest.java]: ok, in this case of course no special importer test is needed. sorry for the confusion 🍡

**Pull request merged**